juju_verify/utils/unit.py

```diff
@@ -166,13 +196,39 @@ def get_applications_names(model: Model, application: str) -> List[str]:
     return applications
 
 
+def get_related_charm_units_to_app(application: Application, charm: str) -> List[Unit]:
+    """Get all units for the same charm related to application.
+
+    :param application: Juju application
+    :param charm: charm name, e.g. ceph-osd
+    """
+    return [
+        unit
+        for relation in application.relations
+        for unit in relation.provides.application.units
+        if parse_charm_name(unit.charm_url) == charm
```

Comment:
It seems the application object also has the [`charm_url`](https://github.com/juju/python-libjuju/blob/8d62315acc978b6b0375931440ec9b85bfaa5a49/juju/application.py#L511) property. I understand the use of a list comprehension here (to retrieve all the affected units), but an alternative could be to use 2 different list comprehensions.

The reason to suggest this alternative is that we are looping through all the units of an application when we know that all units use the same charm (it comes inherited from the application object). It would be faster to check the applications (and then retrieve all units from the filtered applications) than checking all the units (related) in a model (e.g. 200 units of 3 applications).

What do you think?

Reply:
> It seems the application object also has the `charm_url` property.

That is good information. I think I will try something like this:
```python
units = []
for relation in application.relations:
  if parse_charm_name(relation.provides.application.charm_url) == charm:
    units.extend(relation.provides.application.units)

return units
```